### PR TITLE
added productMap and buildUuid instrumentation

### PR DESCRIPTION
**Pull request by dhavalshah24s**
Added extra productmap and buildUuid capabilitites

### CODE REVIEW

File: nightwatch/globals.js [hunk 1]
```diff
@@ -447,3 +449,29 @@ const cucumberPatcher = () => {
     Logger.debug(`Error while patching cucumber ${error}`);
   }
 };
+
+const addProductMapAndbuildUuidCapability = (settings) => {
```
Comment (karanshah-browserstack) [inline nightwatch/globals.js:453]: Add inside try catch block, don't want session to be failing for any unknown exceptions inside our functions.

**Review by karanshah-browserstack — APPROVED** [review]

File: src/testObservability.js [hunk 2]
```diff
@@ -14,7 +14,9 @@ class TestObservability {
   configure(settings = {}) {
     this._settings = settings['@nightwatch/browserstack'] || {};
 
-    if (this._settings.test_observability) {
+    process.env.BROWSERSTACK_TEST_OBSERVABILITY = true;
+
+    if (this._settings.test_observability && this._settings.test_observability.enabled !== undefined) {
```
Comment (amaanbs) [inline src/testObservability.js:19]: This won't work if test_observability is set to false. We'll still end up setting the environment variable to true.

File: nightwatch/globals.js [hunk 3]
```diff
@@ -447,3 +449,33 @@ const cucumberPatcher = () => {
...
+  try {
+    if (!settings?.desiredCapabilities?.['bstack:options']) {
```
Comment (amaanbs) [inline nightwatch/globals.js:455]: Are we not handling non w3c mode?

File: nightwatch/globals.js [hunk 4]
```diff
@@ -447,3 +449,33 @@ const cucumberPatcher = () => {
...
+    } else {
+      automate = true;
+    }
+    
+    const buildProductMap = {
+      automate: automate,
+      app_automate: app_automate,
+      observability: helper.isTestObservabilitySession(),
+      accessibility: helper.isAccessibilitySession(),
+      turboscale: false,
```
Comment (amaanbs) [inline nightwatch/globals.js:472]: We need to track for turboscale as well.
https://github.com/search?q=repo%3Anightwatchjs%2Fnightwatch-plugin-browserstack%20turboscale&type=code

**Review by amaanbs — CHANGES_REQUESTED** [review]

File: nightwatch/globals.js [hunk 5]
```diff
@@ -452,29 +452,26 @@ const cucumberPatcher = () => {
...
-    if (settings.desiredCapabilities['appUploadPath'] || settings.desiredCapabilities['appUploadUrl'] || settings.desiredCapabilities['app'] || settings.desiredCapabilities['appium:options']) {
-      app_automate = true;
-    } else {
-      automate = true;
-    }
-    
     const buildProductMap = {
-      automate: automate,
-      app_automate: app_automate,
+      automate: helper.getObservabilityLinkedProductName() === 'automate',
```
Comment (amaanbs) [inline nightwatch/globals.js:460]: Redundant method calls. Refactor.

File: nightwatch/globals.js [hunk 6]
```diff
@@ -452,29 +452,26 @@ const cucumberPatcher = () => {
...
       percy: false
     };
     
-    settings.desiredCapabilities['bstack:options']['buildProductMap'] = buildProductMap;
-    settings.desiredCapabilities['bstack:options']['testhubBuildUuid'] = process.env.BS_TESTOPS_BUILD_HASHED_ID ? process.env.BS_TESTOPS_BUILD_HASHED_ID : '' ;
+    if (settings.desiredCapabilities['bstack:options']) {
+      settings.desiredCapabilities['bstack:options']['buildProductMap'] = buildProductMap;
+      settings.desiredCapabilities['bstack:options']['testhubBuildUuid'] = process.env.BS_TESTOPS_BUILD_HASHED_ID ? process.env.BS_TESTOPS_BUILD_HASHED_ID : '' ;
+    } else {
+      settings.desiredCapabilities['buildProductMap'] = buildProductMap;
```
Comment (amaanbs) [inline nightwatch/globals.js:472]: Does this need a prefix of `browserstack.` ? Is this reflected on the dashboard correctly?

**Review by amaanbs — CHANGES_REQUESTED** [review]

**Review by karanshah-browserstack — APPROVED** [review]

**Review by 07souravkunda — APPROVED** [review]

**Review by garg3133 — APPROVED** [review]

**Pull request merged**